rskj-core/src/main/java/co/rsk/net/BlockSyncService.java

```diff
@@ -0,0 +1,332 @@
+/*
+ * This file is part of RskJ
+ * Copyright (C) 2017 RSK Labs Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package co.rsk.net;
+
+import co.rsk.core.bc.BlockChainStatus;
+import co.rsk.core.bc.BlockUtils;
+import co.rsk.net.messages.*;
+import org.ethereum.core.*;
+import org.ethereum.db.ByteArrayWrapper;
+import org.ethereum.net.server.ChannelManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.spongycastle.util.encoders.Hex;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+import java.math.BigInteger;
+import java.util.*;
+
+/**
+ * BlockSyncService processes blocks to add into a blockchain.
+ * If a block is not ready to be added to the blockchain, it will be on hold in a BlockStore.
+ * <p>
+ * This class is tightly coupled with
+ */
+public class BlockSyncService {
```

Comment:
This class is ugly as hell, lets talk about it later

Reply:
Agreed. But this is just pieces of `NodeBlockProcessor` that were pulled out and are now shared with `SyncProcessor`. You can take a look at the original: https://github.com/rsksmart/rskj/blob/9a90e362d553ae885a7fa1902f6d4f49dbdb870a/rskj-core/src/main/java/co/rsk/net/NodeBlockProcessor.java

We didn't work on improving this particular piece of code, but on reusability of working code. At least now we have two smaller, more manageable classes!